### ISSUE TITLE
Current master does not build under -fno-exceptions

### ISSUE THREAD

**Issue by dylanhart** [capnproto/capnproto#812]
There is an unguarded `catch` in kj/async-io.c++. I haven't tried compiling past this. `v0.7.0` compiles fine.

Additionally, would you consider adding a `CAPNP_NOEXCEPT` cmake option? I know that -fno-exceptions is only partially supported, but it would be nice to not have to patch this in.

Thanks,
Dylan

---

#### System
* Alpine Linux v3.8.4
* gcc (Alpine 6.4.0) 6.4.0
* cmake 3.11.1

#### make --verbose output:

```
cd /git/lib-cpp-capnproto/build/x86_64-linux/debug/src/kj && /usr/bin/c++   -I/git/lib-cpp-capnproto/capnproto/c++/src/kj/..  -g   -D KJ_HAS_ZLIB=1 -fno-exceptions -Wall -Wextra -Wno-strict-aliasing -Wno-sign-compare -Wno-unused-parameter -pthread -std=gnu++14 -o CMakeFiles/kj-async.dir/async-io.c++.o -c /git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++: In member function 'virtual kj::Promise<void> kj::{anonymous}::AsyncTee::PumpSink::fill(kj::{anonymous}::AsyncTee::Buffer&, const kj::Maybe<kj::OneOf<kj::{anonymous}::AsyncTee::Eof, kj::Exception> >&)':
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:1402:35: error: exception handling disabled, use -fexceptions to enable
         } catch (const Exception& exception) {
                                   ^~~~~~~~~
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:1403:21: error: 'exception' was not declared in this scope
           reject(cp(exception));
                     ^~~~~~~~~
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:1403:21: note: suggested alternative:
In file included from /usr/include/c++/6.4.0/new:40:0,
                 from /usr/include/c++/6.4.0/ext/new_allocator.h:33,
                 from /usr/include/c++/6.4.0/x86_64-alpine-linux-musl/bits/c++allocator.h:33,
                 from /usr/include/c++/6.4.0/bits/allocator.h:46,
                 from /usr/include/c++/6.4.0/deque:61,
                 from /git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:34:
/usr/include/c++/6.4.0/exception:60:9: note:   'std::exception'
   class exception
         ^~~~~~~~~
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++: In lambda function:
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:1606:32: error: 'exception' was not declared in this scope
         stoppage = Stoppage(cp(exception));
                                ^~~~~~~~~
/git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:1606:32: note: suggested alternative:
In file included from /usr/include/c++/6.4.0/new:40:0,
                 from /usr/include/c++/6.4.0/ext/new_allocator.h:33,
                 from /usr/include/c++/6.4.0/x86_64-alpine-linux-musl/bits/c++allocator.h:33,
                 from /usr/include/c++/6.4.0/bits/allocator.h:46,
                 from /usr/include/c++/6.4.0/deque:61,
                 from /git/lib-cpp-capnproto/capnproto/c++/src/kj/async-io.c++:34:
/usr/include/c++/6.4.0/exception:60:9: note:   'std::exception'
   class exception
         ^~~~~~~~~
```

**Comment by kentonv** [comment]
Sorry, we only test and fix exotic configurations like `-fno-exceptions` when we do a release. These configurations are not maintained at git master, and commonly break between releases as it is rather cumbersome for us to test every commit with every combination of flags. :/

We generally recommend using a release build, and especially so if you depend on support for `-fno-exceptions`, `-fno-rtti`, Android, etc.

Regarding a `CAPNP_NOEXCEPT` cmake option, how would this differ from setting `CMAKE_CXX_FLAGS=-fno-exceptions`?

**Comment by dylanhart** [comment]
Thanks, for the update. It turns out that I was setting CMAKE_CXX_FLAGS wrong in my testing, and I CMakeLists.txt to add add_compiler_flags(-fno-exceptions) as a workaround. I'll close this issue as it's a combination of not supported and user error.

**Comment by harrishancock** [comment]
I have a PR up to fix this (#781), but ran into issues on Windows. I'll revisit it before the next release.
